subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c

```diff
@@ -674,97 +690,123 @@ static int firmware_block_received_cb(uint16_t obj_inst_id, uint16_t res_id, uin
 				      uint8_t *data, uint16_t data_len, bool last_block,
 				      size_t total_size)
 {
-	uint8_t curent_percent;
-	uint32_t current_bytes;
-	size_t offset;
-	size_t skip = 0;
-	int ret = 0;
-	int image_type;
+	enum dfu_target_image_type image_type;
 
 	if (!data_len) {
+		k_sem_give(&fw_buf_sem);
 		return -EINVAL;
 	}
+	if (ongoing_obj_id != UNUSED_OBJ_ID && ongoing_obj_id != obj_inst_id) {
+		LOG_INF("DFU is allocated already");
+		k_sem_give(&fw_buf_sem);
+		return -EAGAIN;
+	}
 
-	if (bytes_downloaded == 0) {
-		if (ongoing_obj_id != UNUSED_OBJ_ID) {
-			LOG_INF("DFU is allocated already");
-			return -EAGAIN;
-		}
-
-		ongoing_obj_id = obj_inst_id;
-		client_acknowledge();
-
+	if (push_data.bytes_downloaded == 0) {
 		image_type = dfu_target_img_type(data, data_len);
 		if (image_type == DFU_TARGET_IMAGE_TYPE_NONE) {
-			ret = -ENOMSG; /* Translates to unsupported image type */
-			goto cleanup;
+			k_sem_give(&fw_buf_sem);
+			return -ENOMSG; /* Translates to unsupported image type */
 		}
 		LOG_INF("Image type %d", image_type);
-		ret = fota_download_util_dfu_target_init(image_type);
+		push_data = (struct push_data){
+			.work = push_data.work,
+			.total_size = total_size,
+			.image_type = image_type,
+			.data = data,
+			.data_len = data_len,
+			.last_block = last_block,
+		};
+	} else {
+		if (push_data.return_code) {
+			/* If previous write was not succesfull, do not continue */
+			k_sem_give(&fw_buf_sem);
+			return push_data.return_code;
```

Comment:
Does this mean that we still continue to receive blocks? Should we?

Reply:
No, when we return the error code from block-write callback, the LwM2M engine return error to the CoAP Block-Wise transfer, usually something like "Internal error 500" so it stops the server from sending more blocks.